src/toxcore/dht/client.rs

```diff
@@ -0,0 +1,155 @@
+/*
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Manage ping_id.
+Generate ping_id on request packet, check ping_id on response packet.
+*/
+
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+use toxcore::crypto_core::*;
+
+/// peer info.
+#[derive(Clone, Debug)]
+pub struct ClientData {
+    /// hash of ping_ids to check PingResponse is correct
+    ping_hash: HashMap<u64, Instant>,
+    /// last received ping-response time
+    pub last_resp_time: Instant
+}
+
+impl ClientData {
+    /// create ClientData object
+    pub fn new() -> ClientData {
+        ClientData {
+            ping_hash: HashMap::new(),
+            last_resp_time: Instant::now(),
+        }
+    }
+    /// set new random ping id to the client and return it
+    fn new_ping_id(&mut self) -> u64 {
+        loop {
+            let ping_id = random_u64();
+            if ping_id != 0 && !self.ping_hash.contains_key(&ping_id) {
+                return ping_id;
+            }
+        }
+    }
+
+    /// clear timed out ping_id
+    pub fn clear_timedout(&mut self, timeout: Duration) {
+        self.ping_hash.retain(|&_ping_id, &mut time|
+            time.elapsed() <= timeout);
+    }
+
+    /// Add a Ping Hash Entry and return a new ping_id.
+    pub fn add_ping_id(&mut self) -> u64 {
+        let ping_id = self.new_ping_id();
+        self.ping_hash.insert(ping_id, Instant::now());
+
+        ping_id
+    }
+
+    /// Check if ping_id is valid and not timed out.
+    pub fn is_valid_ping_id(&mut self, ping_id: u64, timeout: Duration) -> bool {
+        if ping_id == 0 {
+            error!("Given ping_id is 0");
+            return false
+        }
+
+        let time_ping_sent = match self.ping_hash.remove(&ping_id) {
```

Comment:
Function with name `is_valid_ping_id` must not remove anything.

Reply:
That's right.